### PR TITLE
Accept either a ranking from candidate_id to rank or an order list of…

### PR DESCRIPTION
**Pull request by jessebutterfield**
… candidate ids in the voting urls

### CODE REVIEW

**Comment by jessebutterfield** [comment]
@DSASanFrancisco/portal-members

File: membership/web/elections.py [hunk 1]
```diff
@@ -113,6 +114,15 @@ def submit_paper_vote(requester: Member, session: Session):
     if election.status == 'final':
         return BadRequest('You may not submit more votes after an election has been marked final')
     vote_key = request.json['ballot_key']
+
+    rankings = request.json.get('rankings')
+    if isinstance(rankings, dict):
+        rankings = sorted(rankings.items(), key=operator.itemgetter(1))
+        rankings = [ranking[0] for ranking in rankings]
+    if not isinstance(rankings, list):
+        return BadRequest('Rankings must either a list of candidate ids or a map'
```
Comment (jeffmay) [inline membership/web/elections.py]: `must either a list`

File: membership/web/elections.py [hunk 2]
```diff
@@ -113,6 +114,15 @@ def submit_paper_vote(requester: Member, session: Session):
     if election.status == 'final':
         return BadRequest('You may not submit more votes after an election has been marked final')
     vote_key = request.json['ballot_key']
+
+    rankings = request.json.get('rankings')
+    if isinstance(rankings, dict):
+        rankings = sorted(rankings.items(), key=operator.itemgetter(1))
```
Comment (hyena) [inline membership/web/elections.py]: Why not just `key=lambda x: x[1]`? No big deal. I think these two lines could actually be:
`rankings = sorted(rankings.keys(), key=lambda x: rankings[x])`

We should also have some sanity checking of input here:
No candidate should appear twice. Ranks should be contiguous numbers.

File: membership/web/elections.py [hunk 3]
```diff
@@ -145,6 +157,15 @@ def submit_vote(requester: Member, session: Session):
     election = session.query(Election).get(election_id)
     if election.status == 'final' or election.status == 'polls closed':
         return BadRequest('You may not submit a vote after the polls have closed')
+
+    rankings = request.json.get('rankings')
+    if isinstance(rankings, dict):
+        rankings = sorted(rankings.items(), key=operator.itemgetter(1))
```
Comment (jeffmay) [inline membership/web/elections.py]: TIL `operator.itemgetter(1)` creates a function that retrieves the second value of a sequence.

Comment (jeffmay) [inline membership/web/elections.py]: Today I also learned (TIAL) that Python 3.6 dictionaries are ordered by default: https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-compactdict

File: membership/web/elections.py [hunk 4]
```diff
@@ -145,6 +157,15 @@ def submit_vote(requester: Member, session: Session):
...
+        rankings = [ranking[0] for ranking in rankings]
+    if not isinstance(rankings, list):
+        return BadRequest('Rankings must either a list of candidate ids or a map'
```
Comment (jeffmay) [inline membership/web/elections.py]: Same typo here.

File: membership/web/elections.py (hunk 2 above)
Comment (jessebutterfield) [inline membership/web/elections.py]: The election actually doesn't require either of those constraints. We should make sure they are candidates for the right election. Do we think that's important enough to add right now? And I like your one-liner. Let's do all that stuff right after we pull this election off.

**Comment by jessebutterfield** [comment]
The main idea was to just make it easier on the front end person so they didn't have to sort if people were entering rankings into a form.

**Comment by jeffmay** [comment]
My take is that the paper ballot should actually be represented in its truest form in the database. If someone enters an invalid ballot, we should put that invalid ballot into the database and handle it on the way out when counting. That way we can always display and audit the original ballot. Right now, we have a bit of a hybrid. It'll work for this election though, I suppose.

File: membership/web/elections.py (hunk 2 above)
Comment (jeffmay) [inline membership/web/elections.py]: Yea, unfortunately, the real world of paper ballots prohibits us from making these constaints. We should serve to upload the ballots as accurately as written. It should be the job of the code to decide what to do when encountering these scenarios.

+1 to the one-liner tho.

**Comment by jessebutterfield** [comment]
We decided the api should only accept one form of request and fixed this issue on the front end instead.

**Pull request closed without merging**